src/Compatibility/Microsoft.DotNet.PackageValidation/ValidatePackage.cs

```diff
@@ -30,30 +33,58 @@ public class ValidatePackage : TaskBase
             }
 
-            Package package = NupkgParser.CreatePackage(PackageTargetPath, runtimeGraph);
-            PackageValidationLogger logger = new(Log, CompatibilitySuppressionFilePath, GenerateCompatibilitySuppressionFile);
-
-            new CompatibleTfmValidator(NoWarn, null, RunApiCompat, EnableStrictModeForCompatibleTfms, logger).Validate(package);
-            new CompatibleFrameworkInPackageValidator(NoWarn, null, EnableStrictModeForCompatibleFrameworksInPackage, logger).Validate(package);
+            AppDomain.CurrentDomain.AssemblyResolve += ResolverForRoslyn;
 
-            if (!DisablePackageBaselineValidation && !string.IsNullOrEmpty(BaselinePackageTargetPath))
+            try
             {
-                Package baselinePackage = NupkgParser.CreatePackage(BaselinePackageTargetPath, runtimeGraph);
-                new BaselinePackageValidator(baselinePackage, NoWarn, null, RunApiCompat, logger).Validate(package);
+                RuntimeGraph runtimeGraph = null;
+                if (!string.IsNullOrEmpty(RuntimeGraph))
+                {
+                    runtimeGraph = JsonRuntimeFormat.ReadRuntimeGraph(RuntimeGraph);
+                }
+
+                Package package = NupkgParser.CreatePackage(PackageTargetPath, runtimeGraph);
+                PackageValidationLogger logger = new(Log, CompatibilitySuppressionFilePath, GenerateCompatibilitySuppressionFile);
+
+                new CompatibleTfmValidator(NoWarn, null, RunApiCompat, EnableStrictModeForCompatibleTfms, logger).Validate(package);
+                new CompatibleFrameworkInPackageValidator(NoWarn, null, EnableStrictModeForCompatibleFrameworksInPackage, logger).Validate(package);
+
+                if (!DisablePackageBaselineValidation && !string.IsNullOrEmpty(BaselinePackageTargetPath))
+                {
+                    Package baselinePackage = NupkgParser.CreatePackage(BaselinePackageTargetPath, runtimeGraph);
+                    new BaselinePackageValidator(baselinePackage, NoWarn, null, RunApiCompat, logger).Validate(package);
+                }
+
+                if (GenerateCompatibilitySuppressionFile)
+                {
+                    logger.GenerateSuppressionsFile(CompatibilitySuppressionFilePath);
+                }
             }
-
-            if (GenerateCompatibilitySuppressionFile)
+            finally
             {
-                logger.GenerateSuppressionsFile(CompatibilitySuppressionFilePath);
+                AppDomain.CurrentDomain.AssemblyResolve -= ResolverForRoslyn;
             }
         }
+
+        private Assembly ResolverForRoslyn(object sender, ResolveEventArgs args)
+        {
+            AssemblyName name = new(args.Name);
+
+            return name.Name switch
+            {
+                "Microsoft.CodeAnalysis" or "Microsoft.CodeAnalysis.CSharp" =>
+                    Assembly.LoadFrom(Path.Combine(RoslynDirectory, $"{name.Name}.dll")),
```

Comment:
We ought to check the version we loaded is sufficient to satisfy our reference.  If it isn't, the runtime will "trust" us and then we'd be downgrading the assembly reference and could run into missing members later on.